library/core/src/intrinsics.rs

```diff
@@ -3503,6 +3528,28 @@ pub const unsafe fn write_bytes<T>(dst: *mut T, val: u8, count: usize) {
     }
 }
 
+/// Return whether the initialization state is preserved.
+///
+/// This is used for contracts only.
+/// FIXME: Change this once we add support to quantifiers.
+#[allow(dead_code)]
+#[allow(unused_variables)]
+fn check_copy_untyped<T>(src: *const T, dst: *mut T, count: usize) -> bool {
+    #[cfg(kani)]
+    if count > 0 {
+        let byte = kani::any_where(|sz: &usize| *sz < size_of::<T>());
+        let elem = kani::any_where(|val: &usize| *val < count);
+        let src_data = src as *const u8;
+        let dst_data = dst.wrapping_add(elem) as *const u8;
+        ub_checks::can_dereference(unsafe { src_data.add(byte) })
+            == ub_checks::can_dereference(unsafe { dst_data.add(byte) })
```

Comment:
Can you please explain to me what this trying to do? I'd read this as just comparing is-a-valid-pointer of `(src as *const u8) + byte` and `((dst + elem) as *const u8) + byte`. Why is is-a-valid-pointer a meaningful test here, and why compare different offsets in src and dst?

Reply:
This is checking that a copied byte is safe to read if and only if the source byte is also safe to read.

Ideally this should be using quantifiers, but we don't support them yet.

Reply:
I get the lack of quantification. But what got me surprised is that you aren't comparing the safe-to-read for the same byte offsets (`dst_data` has that extra `+ elem` offset)?!

Reply:
That's because `src_data` gets copied `count` times into to `dst_data`. I.e., in practice, if `src_data` points to a type `T`, `dst_data` must points to `[T]` of size greater or equal to `count`.

Ideally, this should be something like:
```rust
forall!(byte in 0..size_of::<T>(), {
    forall!(elem in 0..count, {
         can_dereference(src[byte]) == can_dereference(dst[elem][byte])
    })
});
```
Since we don't have quantifiers, we get an arbitrary byte, and an arbitrary copied element in the slice, and we check that the initialization state is kept.

Reply:
Ah, thank you, now I got it.